### ISSUE TITLE
Post-training quantization for efficientnet_lite give bad accuracy 

### ISSUE THREAD

**Issue by HoangTienDuc** [tensorflow/tensorflow#55660]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Performance

### Source

source

### Tensorflow Version

tf 2.8

### Custom Code

No

### OS Platform and Distribution

ubntu 18.04

### Mobile device

None

### Python version

3.8.10

### Bazel version

None

### GCC/Compiler version

None

### CUDA/cuDNN version

None

### GPU model and memory

None

### Current Behaviour?

```shell
Thank for your image classification model [maker](https://www.tensorflow.org/lite/tutorials/model_maker_image_classification). By follow your tutorial, i create a training script with a purpose to use efficientnet-lite as our classifier.

However, after export tflite model (efficientnet_lite2). Without quantization, out model get high accuracy (99.4% on my test dataset), very low speed (1.3748071193695068/image).
So i decide to use quantization to get better speed. 
But the accuray is really bad. 



### Relevant log output

```shell
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
hub_keras_layer_v1v2 (HubKer (None, 1280)              4869168   
_________________________________________________________________
dropout (Dropout)            (None, 1280)              0         
_________________________________________________________________
dense (Dense)                (None, 2)                 2562      
=================================================================
Total params: 4,871,730
Trainable params: 2,562
Non-trainable params: 4,869,168
_________________________________________________________________
None
/usr/local/lib/python3.8/dist-packages/keras/optimizer_v2/optimizer_v2.py:355: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  warnings.warn(
Epoch 1/5
2022-04-19 03:05:06.677678: I tensorflow/stream_executor/cuda/cuda_dnn.cc:381] Loaded cuDNN version 8204
455/455 [==============================] - 204s 442ms/step - loss: 0.2681 - accuracy: 0.9708 - val_loss: 0.2391 - val_accuracy: 0.9897
Epoch 2/5
455/455 [==============================] - 202s 441ms/step - loss: 0.2428 - accuracy: 0.9881 - val_loss: 0.2343 - val_accuracy: 0.9930
Epoch 3/5
455/455 [==============================] - 203s 443ms/step - loss: 0.2396 - accuracy: 0.9901 - val_loss: 0.2320 - val_accuracy: 0.9933
Epoch 4/5
455/455 [==============================] - 203s 445ms/step - loss: 0.2373 - accuracy: 0.9909 - val_loss: 0.2308 - val_accuracy: 0.9941
Epoch 5/5
455/455 [==============================] - 202s 442ms/step - loss: 0.2365 - accuracy: 0.9913 - val_loss: 0.2301 - val_accuracy: 0.9940
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
hub_keras_layer_v1v2 (HubKer (None, 1280)              4869168   
_________________________________________________________________
dropout (Dropout)            (None, 1280)              0         
_________________________________________________________________
dense (Dense)                (None, 2)                 2562      
=================================================================
Total params: 4,871,730
Trainable params: 2,562
Non-trainable params: 4,869,168
_________________________________________________________________
228/228 [==============================] - 33s 54ms/step - loss: 0.2290 - accuracy: 0.9951
############### FOR FP32 MODEL
2022-04-19 03:22:31.386912: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
2022-04-19 03:22:33.971451: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:22:33.972101: I tensorflow/core/grappler/devices.cc:66] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 1
2022-04-19 03:22:33.972400: I tensorflow/core/grappler/clusters/single_machine.cc:357] Starting new session
2022-04-19 03:22:33.973941: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:22:33.974352: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:22:33.974627: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:22:33.976490: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:22:33.976763: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:22:33.977095: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1510] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 8938 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:2d:00.0, compute capability: 7.5
2022-04-19 03:22:34.005010: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:1164] Optimization results for grappler item: graph_to_optimize
  function_optimizer: Graph size after: 1188 nodes (856), 1203 edges (869), time = 13.073ms.
  function_optimizer: function_optimizer did nothing. time = 0.225ms.

2022-04-19 03:22:34.653218: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:351] Ignored output_format.
2022-04-19 03:22:34.653265: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:354] Ignored drop_control_dependency.
2022-04-19 03:22:34.711780: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:210] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
fully_quantize: 0, inference_type: 6, input_inference_type: 3, output_inference_type: 3
WARNING:absl:For model inputs containing unsupported operations which cannot be quantized, the `inference_input_type` attribute will default to the original type.
############### FOR FP16 MODEL
2022-04-19 03:32:07.372624: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:32:07.372880: I tensorflow/core/grappler/devices.cc:66] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 1
2022-04-19 03:32:07.372951: I tensorflow/core/grappler/clusters/single_machine.cc:357] Starting new session
2022-04-19 03:32:07.373226: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:32:07.373494: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:32:07.373724: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:32:07.373987: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:32:07.374212: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-04-19 03:32:07.374402: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1510] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 8938 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:2d:00.0, compute capability: 7.5
2022-04-19 03:32:07.397998: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:1164] Optimization results for grappler item: graph_to_optimize
  function_optimizer: Graph size after: 1188 nodes (856), 1203 edges (869), time = 12.634ms.
  function_optimizer: function_optimizer did nothing. time = 0.21ms.

2022-04-19 03:32:08.278801: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:351] Ignored output_format.
2022-04-19 03:32:08.278847: W tensorflow/compiler/mlir/lite/python/tf_tfl_flatbuffer_helpers.cc:354] Ignored drop_control_dependency.
```
</details>

**Comment by HoangTienDuc** [comment]
Hi everyone, Is there any update?

**Comment by mohantym** [comment]
Hi @HoangTienDuc ! Sorry for the late response. Could you check with the below code snippet and let us know the difference. 
```
import tensorflow as tf
print(tf.__version__)

model_file = '/content/mobilenet_v2_1.0_224_frozen_pb' // your model here
converter = tf.lite.TFLiteConverter.from_saved_model(model_file)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
#converter.target_spec.supported_ops = [
 # tf.lite.OpsSet.TFLITE_BUILTINS, # enable TensorFlow Lite ops.
  #tf.lite.OpsSet.SELECT_TF_OPS # enable TensorFlow ops.
#] use select ops too if you face any select ops issue
converter.experimental_new_converter = True //enables MLIR operator conversion.
converter.allow_custom_ops = True
tflite_model = converter.convert()
```

**Comment by HoangTienDuc** [comment]
Hi @mohantym 
Thank for your response
However, the result is still very bad. 

# Reproduce

### Train code 
```
import os

import numpy as np

import tensorflow as tf
assert tf.__version__.startswith('2')

from tflite_model_maker import model_spec
from tflite_model_maker import image_classifier
from tflite_model_maker.config import ExportFormat
from tflite_model_maker.config import QuantizationConfig
from tflite_model_maker.image_classifier import DataLoader

# PATH TO FOLDER CONTAIN DATASET
image_path = "/ws/dataset/flower_photos_preprocess"

data = DataLoader.from_folder(image_path)

train_data, rest_data = data.split(0.80)
validation_data, test_data = rest_data.split(0.5)


# IMAGE_CLASSIFICATION_MODELS = [
    # 'efficientnet_lite0', 'efficientnet_lite1', 'efficientnet_lite2',
    # 'efficientnet_lite3', 'efficientnet_lite4', 'mobilenet_v2', 'resnet_50'
# ]
# model = image_classifier.create(train_data, model_spec=model_spec.get('mobilenet_v2'), validation_data=validation_data)
model_dir = "/ws/tensorflow/tensorflow/lite/g3doc/tutorials/models/2"
model = image_classifier.create(train_data, model_spec=model_spec.get('efficientnet_lite2'), validation_data=validation_data, \
    batch_size=128, shuffle=True, model_dir=model_dir, epochs=100, use_hub_library=False)

model.summary()

loss, accuracy = model.evaluate(test_data)


print("############### EVELUATE FP32 MODEL")
model.export(export_dir=model_dir, \
    tflite_filename='model.tflite')
# model.evaluate_tflite(os.path.join(model_dir, 'model.tflite'), test_data)

model.export(export_dir=model_dir, export_format=ExportFormat.SAVED_MODEL)
# model.evaluate_tflite('model.tflite', test_data)



print("############### EVELUATE FP16 MODEL")
config = QuantizationConfig.for_float16()
model.export(export_dir=model_dir, \
    tflite_filename='model_fp16.tflite', quantization_config=config)
model.evaluate_tflite(os.path.join(model_dir, 'model_fp16.tflite'), test_data)
```
### Export code
```
import tensorflow as tf
print(tf.__version__)

model_file = '/ws/models/2/saved_model' #// your model here
converter = tf.lite.TFLiteConverter.from_saved_model(model_file)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
#converter.target_spec.supported_ops = [
 # tf.lite.OpsSet.TFLITE_BUILTINS, # enable TensorFlow Lite ops.
  #tf.lite.OpsSet.SELECT_TF_OPS # enable TensorFlow ops.
#] use select ops too if you face any select ops issue
converter.experimental_new_quantizer = True #//enables MLIR operator  quantization
converter.allow_custom_ops = True
tflite_model = converter.convert()
with open('model_fp16_haha.tflite', 'wb') as f:
  f.write(tflite_model)

```
### Test code
```
import tflite_runtime.interpreter as tflite     
import cv2
import numpy as np
import os

class Classifier():
    def __init__(self, model_file="model_fp16_haha.tflite"):
        self.model_file = model_file
    
    def prepare(self):
        self.interpreter = tflite.Interpreter(model_path=self.model_file)
        self.interpreter.allocate_tensors()
        self.input_details = self.interpreter.get_input_details()
        self.output_details = self.interpreter.get_output_details()
        self.width = self.input_details[0]['shape'][1]
        self.height = self.input_details[0]['shape'][2]

    def preprocess(self, img):
        img = cv2.resize(img, (self.width, self.height))
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        img = (img - 128.) / 128.
        input_data = np.expand_dims(img, axis=0).astype('float32')
        return input_data
    def run(self, img):
        input_data = self.preprocess(img)
        self.interpreter.set_tensor(self.input_details[0]['index'], input_data)
        self.interpreter.invoke()
        output_data = self.interpreter.get_tensor(self.output_details[0]['index'])
        results = np.squeeze(output_data)
        top_k = np.argmax(results)
        return top_k



model = Classifier()
model.prepare()

test_dir = "/ws/re-training/dataset/flower_photos/daisy/"
image_names = os.listdir(test_dir)
for image_name in image_names:
    path = os.path.join(test_dir, image_name)
    image = cv2.imread(path)
    print(model.run(image))
```

**Comment by pjpratik** [comment]
@HoangTienDuc I have run the code using tflite model maker nightly version and I have observed that fp16 quantization gives comparable results with non quantized one. Could you please check this [gist](https://colab.research.google.com/gist/pjpratik/db209192ef7aff2d5e9bd278b1d92c51/55660.ipynb) and let us know if it helps. Thank you.
